jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java

```diff
@@ -522,16 +467,92 @@ public String toString()
 
             return String.format("%s{capacity=%d,inuse=%d(%d%%)}",
                 super.toString(),
-                _capacity,
+                getCapacity(),
                 inUse,
                 entries > 0 ? (inUse * 100) / entries : 0);
         }
+
+        private class CompoundBucket extends CompoundPool<RetainableByteBuffer>
+        {
+            private final QueuedBucket queuedBucket;
+
+            private CompoundBucket(ConcurrentBucket concurrentBucket, QueuedBucket queuedBucket)
+            {
+                super(concurrentBucket, queuedBucket);
+                this.queuedBucket = queuedBucket;
+            }
+
+            private void evict(Entry<RetainableByteBuffer> entry)
+            {
+                if (entry instanceof ConcurrentPool.ConcurrentEntry<RetainableByteBuffer> concurrentEntry)
+                {
+                    Entry<RetainableByteBuffer> queuedEntry = queuedBucket.acquire();
+                    if (queuedEntry != null)
+                    {
+                        queuedEntry.remove();
+                        if (!concurrentEntry.release())
+                            concurrentEntry.remove();
+                    }
+                    else
+                    {
+                        concurrentEntry.remove();
+                    }
+                }
+                else
+                {
+                    entry.remove();
+                }
+            }
+        }
+
+        private class ConcurrentBucket extends ConcurrentPool<RetainableByteBuffer>
+        {
+            private ConcurrentBucket(StrategyType strategyType, int maxSize, ToIntFunction<RetainableByteBuffer> maxMultiplex)
+            {
+                super(strategyType, maxSize, maxMultiplex);
+            }
+
+            @Override
+            protected void onAcquired(Entry<RetainableByteBuffer> entry)
+            {
+                super.onAcquired(entry);
+                updateMemory(-_capacity, _direct);
+            }
+
+            @Override
+            protected void onReleased(Entry<RetainableByteBuffer> entry)
+            {
+                super.onReleased(entry);
+                updateMemory(_capacity, _direct);
+            }
```

Comment:
I prefer updating the memory directly in the `acquire`, `reserve` and `release` methods rather than doing it with these side effects on the various pools themselves.    Putting the updates there makes it much clearer how the memory accounting relates to those key operations